JitsiAuthConnection.js

```diff
@@ -0,0 +1,162 @@
+     * error set).
+     */
+    authenticateAndUpgradeRole({
+        // 1. Log the specified XMPP user in.
+        id,
+        password,
+
+        // 2. Let the API client/consumer know as soon as the XMPP user has been
+        //    successfully logged in.
+        onLoginSuccessful,
+
+        // 3. Join the MUC.
+        roomPassword
+    }) {
+        if (this._promise) {
+            return this._promise;
+        }
+
+        this._promise = new Promise((resolve, reject) => {
+            const connectionEstablished = () => {
+                if (this.canceled) {
+                    return;
+                }
+
+                // Let the caller know that the XMPP login was successful.
+                onLoginSuccessful && onLoginSuccessful();
+
+                // Now authenticate with Jicofo and get a new session ID.
+                this.room
+                    = this.xmpp.createRoom(
+                        this.conference.options.name,
+                        this.conference.options.config);
+                this.room.moderator.authenticate()
+                    .then(() => {
+                        if (this.canceled) {
+                            return;
+                        }
+
+                        this.xmpp.disconnect();
+
+                        // At this point we should have the new session ID
+                        // stored in the settings. Jicofo will allow to join the
+                        // room.
+                        this.conference.join(roomPassword);
+
+                        resolve();
+                    })
+                    .catch(({ error, message }) => {
+                        this.xmpp.disconnect();
+
+                        reject({
+                            authenticationError: error,
+                            message
+                        });
+                    });
+            };
+
+            this.xmpp.addListener(
+                CONNECTION_ESTABLISHED,
+                connectionEstablished);
+            this.xmpp.addListener(
+                CONNECTION_FAILED,
+                (connectionError, message) => reject({
+                    connectionError,
+                    message
+                }));
+            this.xmpp.addListener(
+                CONNECTION_DISCONNECTED,
+                () => {
+                    if (this.canceled) {
+                        reject({});
+                    }
+                });
+
+            if (this.canceled) {
+                reject({});
+            } else {
+                this.xmpp.connect(id, password);
+            }
+        });
+
+        return this._promise;
+    }
+
+    /**
+     * Cancels the authentication if it's currently in progress. The promise
+     * returned by {@link authenticateAndUpgradeRole} will be rejected with an
+     * empty object (i.e. none of the error fields set).
+     */
+    cancel() {
+        this.canceled = true;
+        this.xmpp.disconnect();
```

Comment:
Like @saghul said, what if this.xmpp is not connected at all? The documentation says that, if `cancel` is called, the `Promise` will be rejected with an empty object but... What if I call `cancel` before the event `CONFERENCE_FAILED` arrives or before the `catch` of `this.room.moderator.authenticate()` executes?